### ISSUE TITLE
na.rm argument in survey_mean not behaving as in srvyr v. 0.3.8 or in survey::svymean

### ISSUE THREAD

**Issue by zackarno** [gergness/srvyr#106]
It seems the na.rm argument is not working like it did in 0.3.8 or in the most current survey package.  Below is a reprex showing the behavior of srvyr 0.4.0 along with that of the current survey package. I can assure you  that srvyr 0.3.8 produced the same results as the survey package (I just do not know how to produce a reprex using two versions of the same package). Is this by design or by accident? 

```
library(srvyr)
library(survey)
data<-data.frame(I.HB_LEARNING.ratio_schools = c("high",NA,NA,
                                                      "high",NA,"high",NA,"high",
                                                      NA,"high",NA,"high",
                                                      NA,NA,"high",NA,NA,
                                                      "high","high",NA,"high",
                                                      "high",NA,NA,NA,NA,NA,
                                                      NA,NA,"high","high",
                                                      NA,"high","high","high",
                                                      "high",NA,"low",NA,
                                                      "high",NA,NA,"high",NA,
                                                      "low","high",NA,NA,
                                                      NA,"high","high",NA,NA,
                                                      NA,NA,NA,"high",NA,
                                                      NA,NA,"high",NA,NA,NA,
                                                      "high",NA,"high",NA,
                                                      "high","high","high",
                                                      "high","high",NA,NA,
                                                      "high","high","high",NA,NA,
                                                      "high",NA,"high",NA,
                                                      "high",NA,"high",NA,NA,
                                                      "high","high","high",
                                                      "ideal","high",NA,NA,NA,
                                                      "high",NA,NA)
                    )

data$I.HB_LEARNING.ratio_schools<-as.factor(data$I.HB_LEARNING.ratio_schools)
data_srvyr<- as_survey(data)

data_srvyr %>% 
  group_by(I.HB_LEARNING.ratio_schools) %>% 
  summarise(mean_pct=survey_mean(na.rm=T))
data_srvyr %>% 
  group_by(I.HB_LEARNING.ratio_schools) %>% 
  summarise(mean_pct=survey_mean(na.rm=F))

survey::svymean(x = ~I.HB_LEARNING.ratio_schools,design= data_srvyr, na.rm=T)
survey::svymean(x = ~I.HB_LEARNING.ratio_schools,design= data_srvyr, na.rm=F) 
```

**Comment by gergness** [comment]
Hm, yea, this is a little confusing, but I think it's intentional. It's a place where the `survey` package and `dplyr` differ in how they treat group summaries, and we've chosen to be more consistent with `dplyr`.

To get the calculation you want, you should `filter` out the NAs rather than using `na.rm`. 

``` r
library(srvyr)
library(survey)

data<-data.frame(I.HB_LEARNING.ratio_schools = c("high",NA,NA,
                                                 "high",NA,"high",NA,"high",
                                                 NA,"high",NA,"high",
                                                 NA,NA,"high",NA,NA,
                                                 "high","high",NA,"high",
                                                 "high",NA,NA,NA,NA,NA,
                                                 NA,NA,"high","high",
                                                 NA,"high","high","high",
                                                 "high",NA,"low",NA,
                                                 "high",NA,NA,"high",NA,
                                                 "low","high",NA,NA,
                                                 NA,"high","high",NA,NA,
                                                 NA,NA,NA,"high",NA,
                                                 NA,NA,"high",NA,NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high","high","high",
                                                 "high","high",NA,NA,
                                                 "high","high","high",NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high",NA,"high",NA,NA,
                                                 "high","high","high",
                                                 "ideal","high",NA,NA,NA,
                                                 "high",NA,NA)
)

data$I.HB_LEARNING.ratio_schools<-as.factor(data$I.HB_LEARNING.ratio_schools)
data_srvyr<- as_survey(data)



# This is the calculation you're after
data_srvyr %>% 
    filter(!is.na(I.HB_LEARNING.ratio_schools)) %>%
    group_by(I.HB_LEARNING.ratio_schools) %>% 
    summarise(mean_pct=survey_mean())
#> # A tibble: 3 x 3
#>   I.HB_LEARNING.ratio_schools mean_pct mean_pct_se
#>   <fct>                          <dbl>       <dbl>
#> 1 high                          0.935       0.0366
#> 2 ideal                         0.0217      0.0216
#> 3 low                           0.0435      0.0302

survey::svymean(x = ~I.HB_LEARNING.ratio_schools,design= data_srvyr, na.rm=T)
#>                                      mean     SE
#> I.HB_LEARNING.ratio_schoolshigh  0.934783 0.0366
#> I.HB_LEARNING.ratio_schoolsideal 0.021739 0.0216
#> I.HB_LEARNING.ratio_schoolslow   0.043478 0.0302


# srvyr's default behavior is to include NAs as a group like `dplyr` would.
data_srvyr %>% 
    group_by(I.HB_LEARNING.ratio_schools) %>% 
    summarise(mean_pct=survey_mean())
#> # A tibble: 4 x 3
#>   I.HB_LEARNING.ratio_schools mean_pct mean_pct_se
#>   <fct>                          <dbl>       <dbl>
#> 1 high                            0.43      0.0498
#> 2 ideal                           0.01      0.0100
#> 3 low                             0.02      0.0141
#> 4 <NA>                            0.54      0.0501


# I don't really see how this is useful, so don't mind that it's not possible to calculate in srvyr
survey::svymean(x = ~I.HB_LEARNING.ratio_schools,design= data_srvyr, na.rm=F) 
#>                                  mean SE
#> I.HB_LEARNING.ratio_schoolshigh    NA NA
#> I.HB_LEARNING.ratio_schoolsideal   NA NA
#> I.HB_LEARNING.ratio_schoolslow     NA NA
```

**Comment by zackarno** [comment]
Thanks for your example!  To me this kind of seems less consistent with dplyr and now we are left with a na.rm argument that does not actually do anything which may cause issues. According to most data analysis frameworks that I am familiar with when you try to take a mean of a vector with NAs your result should be NA unless na.rm=T is specified. Below I have provided and example which tries to illustrate why I think this is less consistent and potentially problematic.

#> add this code just after your example above

library(tidyverse )
df<- data_srvyr$variables

#109  just adding a numeric variable with NAs for example purposes
df<-df %>% 
  mutate(
    I.HB_LEARNING.ratio_schools= as.character(I.HB_LEARNING.ratio_schools),
    var_dummy= case_when(I.HB_LEARNING.ratio_schools=="high"~3,
              I.HB_LEARNING.ratio_schools=="ideal"~2,
              I.HB_LEARNING.ratio_schools=="low"~1,
              TRUE~NA_real_)
  )


# by taking the mean inside a dplyr piped mutate or summary function the na.rm is an important # argument. Specifying F will return NA 

df %>% 
  summarise(
    mean_var_dummy = mean(var_dummy,na.rm=F)
  )

#> A tibble: 1 x 1
# > mean_var_dummy
# >         <dbl>
#  >           NA

# whereas specifying TRUE will return the mean with NAs removed.
df %>% 
  summarise(
    mean_var_dummy = mean(var_dummy,na.rm=T)
  )

#> A tibble: 1 x 1
#> mean_var_dummy
#>         <dbl>
#>            2.89

I think to be consistent with dplyr it syntax it would make sense to follow these rules, no? 

You also probably have reasons for not doing this, but I also think that in order to better follow dplyr syntax it might also make sense to always put the variables you want to take the mean/percent mean of inside of the survey_mean  function and leave the aggregating/grouping variables inside the group_by function.

**Comment by gergness** [comment]
I wrote this response before seeing you deleted yours, so I’m hoping you figured it out, but just in case I’ll still post. 

This is consistent with dplyr because we’re using the variable as a grouping variable and all functions return a value per group, including NA as a group. The na.rm argument applies only when a variable is specified inside of `survey_mean()`.

Part of why I think this is confusing is that `survey_mean()` has two meanings depending on whether you specify a variable or not. With a variable it’s similar to `mean()`, but without one it’s more similar to `n() / nrow(.)`. I kind of regret this now, but it seemed to clever to me when I started.

**Comment by zackarno** [comment]
Hi sorry, I had deleted my reply because I didn't render how I wanted. 

Thanks for your example! I am reposting and hopefully it will be more clear. I understand your work around with the filter step which can be a fine solution, but I am not sure I understand entirely why you think its better to do that than make the na.rm argument work as before. To me it seems like having the na.rm argument is more in-line with tidyverse/dplyr syntax as this is well integrated into the dplyr chains for both grouped and ungrouped summary and mutate operations. I think that when you take the mean of a vector with NAs present the answer should be NA. With 0.4.0 update it seems strange that there is na.rm argument that treats NAs differently than svymean, mean, max, min (or most other core functions). I think adding this back in as it was in 0.3.8 makes the srvyr package much more powerful, flexible, and easier to wrap into other functions. I also imagine that this was considered when designing the survey package, because it seems unconventional to treat NA the same as any other factor level when calculating the percent means for the different factors levels.

 It does seem a bit tricky given that both survey_mean and svymean operate differently on numeric and factor variables. I'm sure there are reasons you have not done this, but to me the simplest way to make the survey_mean function even more intuitive and dplyresque would be to make it so you always have to put the variable of interest inside the survey_mean function (for both factors and numerics) and you just leave the grouping variables (essentially aggregation levels) outside of the summarise in the group_by function, I think this would also make these functions easier to use and potentially help make it so the na.rm argument operates the same way on factors, numerics, and as other R functions. While this could be a bad idea for many reasons I do not know, either way I feel like the na.rm argument functioning inside survey_mean as before is beneficial.

library(srvyr)
library(survey)
library(tidyverse)

data<-data.frame(I.HB_LEARNING.ratio_schools = c("high",NA,NA,
                                                 "high",NA,"high",NA,"high",
                                                 NA,"high",NA,"high",
                                                 NA,NA,"high",NA,NA,
                                                 "high","high",NA,"high",
                                                 "high",NA,NA,NA,NA,NA,
                                                 NA,NA,"high","high",
                                                 NA,"high","high","high",
                                                 "high",NA,"low",NA,
                                                 "high",NA,NA,"high",NA,
                                                 "low","high",NA,NA,
                                                 NA,"high","high",NA,NA,
                                                 NA,NA,NA,"high",NA,
                                                 NA,NA,"high",NA,NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high","high","high",
                                                 "high","high",NA,NA,
                                                 "high","high","high",NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high",NA,"high",NA,NA,
                                                 "high","high","high",
                                                 "ideal","high",NA,NA,NA,
                                                 "high",NA,NA)
)


# just making a numerical dummy variable to illustrate point on
data<-data %>% 
  mutate(
    I.HB_LEARNING.ratio_schools= as.character(I.HB_LEARNING.ratio_schools),
    var_dummy= case_when(I.HB_LEARNING.ratio_schools=="high"~3,
                         I.HB_LEARNING.ratio_schools=="ideal"~2,
                         I.HB_LEARNING.ratio_schools=="low"~1,
                         TRUE~NA_real_)
  )


# when using dplyr summarise or mutate verbs you do not have filter NAs, you can just set na.rm=T
data %>% 
  summarise(
    mean_var_dummy = mean(var_dummy,na.rm=T)
  )
#>   mean_var_dummy
#> 1       2.891304



# na.rm is well integrated into dplyr summarise/mutate steps.
# if you take the mean of a vector containing NAs with na.rm=F, the result should be NA.
# This does not occur in the updated srvyr package.
# Keeping this argument functional in the srvyr pacakge would make the functions more powerful and useful especially when 
# they are wrapped into other functions for a large variety of reasons.

data %>% 
  summarise(
    mean_var_dummy = mean(var_dummy,na.rm=F)
  )
#>   mean_var_dummy
#> 1             NA

**Comment by zackarno** [comment]
Hi sorry, I had deleted my reply because I didn't render how I wanted.

Thanks for your example! I am reposting and hopefully it will be more clear. I understand your work around with the filter step which can be a fine solution, but I am not sure I understand entirely why you think its better to do that than make the na.rm argument work as before. To me it seems like having the na.rm argument is more in-line with tidyverse/dplyr syntax as this is well integrated into the dplyr chains for both grouped and ungrouped summary and mutate operations. I think that when you take the mean of a vector with NAs present the answer should be NA. With 0.4.0 update it seems strange that there is na.rm argument that treats NAs differently than svymean, mean, max, min (or most other core functions). I think adding this back in as it was in 0.3.8 makes the srvyr package much more powerful, flexible, and easier to wrap into other functions. I also imagine that this was considered when designing the survey package, because it seems unconventional to treat NA the same as any other factor level when calculating the percent means for the different factors levels.

It does seem a bit tricky given that both survey_mean and svymean operate differently on numeric and factor variables. I'm sure there are reasons you have not done this, but to me the simplest way to make the survey_mean function even more intuitive and dplyresque would be to make it so you always have to put the variable of interest inside the survey_mean function (for both factors and numerics) and you just leave the grouping variables (essentially aggregation levels) outside of the summarise in the group_by function, I think this would also make these functions easier to use and potentially help make it so the na.rm argument operates the same way on factors, numerics, and as other R functions. While this could be a bad idea for many reasons I do not know, either way I feel like the na.rm argument functioning inside survey_mean as before is beneficial.

**Comment by gergness** [comment]
Ah, gotcha. GitHub uses [GitHub flavored markdown](https://guides.github.com/features/mastering-markdown/) so you need to wrap code in triple back ticks (```) to format it. Also the [reprex package](https://reprex.tidyverse.org) is really handy for putting R code and output.


The problem is that the variable of interest lives in the `group_by()`, not the `survey_mean()`, that's why I think it's consistent to not have `na.rm` apply here. 

I can see an argument for there being a `na.rm` like argument in `group_by()`, but given how little time I have to work on `srvyr` right now, I can't imagine implementing that in `srvyr`, that would need to be a feature in `dplyr`.

``` r
suppressPackageStartupMessages({
    library(survey)
    library(srvyr)
})
data<-data.frame(I.HB_LEARNING.ratio_schools = c("high",NA,NA,
                                                 "high",NA,"high",NA,"high",
                                                 NA,"high",NA,"high",
                                                 NA,NA,"high",NA,NA,
                                                 "high","high",NA,"high",
                                                 "high",NA,NA,NA,NA,NA,
                                                 NA,NA,"high","high",
                                                 NA,"high","high","high",
                                                 "high",NA,"low",NA,
                                                 "high",NA,NA,"high",NA,
                                                 "low","high",NA,NA,
                                                 NA,"high","high",NA,NA,
                                                 NA,NA,NA,"high",NA,
                                                 NA,NA,"high",NA,NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high","high","high",
                                                 "high","high",NA,NA,
                                                 "high","high","high",NA,NA,
                                                 "high",NA,"high",NA,
                                                 "high",NA,"high",NA,NA,
                                                 "high","high","high",
                                                 "ideal","high",NA,NA,NA,
                                                 "high",NA,NA)
)

data$I.HB_LEARNING.ratio_schools<-as.factor(data$I.HB_LEARNING.ratio_schools)

# Add a numeric variable
set.seed(2020-11-22)
data$numeric_var<-sample(c(0:10, NA), nrow(data), replace = TRUE)


data_srvyr<- as_survey(data)

# Calculating proportion in a group for srvyr
data_srvyr %>%
    group_by(I.HB_LEARNING.ratio_schools) %>%
    summarize(x = survey_mean())
#> # A tibble: 4 x 3
#>   I.HB_LEARNING.ratio_schools     x   x_se
#>   <fct>                       <dbl>  <dbl>
#> 1 high                         0.43 0.0498
#> 2 ideal                        0.01 0.0100
#> 3 low                          0.02 0.0141
#> 4 <NA>                         0.54 0.0501

# dplyr equivalent
data_srvyr$variables %>% 
    group_by(I.HB_LEARNING.ratio_schools) %>%
    summarize(x = n() / nrow(.))
#> `summarise()` ungrouping output (override with `.groups` argument)
#> # A tibble: 4 x 2
#>   I.HB_LEARNING.ratio_schools     x
#>   <fct>                       <dbl>
#> 1 high                         0.43
#> 2 ideal                        0.01
#> 3 low                          0.02
#> 4 <NA>                         0.54


# Calculating mean of a numeric variable (na.rm does apply)
data_srvyr %>%
    summarize(x = survey_mean(numeric_var))
#>    x x_se
#> 1 NA  NaN

data_srvyr %>%
    summarize(x = survey_mean(numeric_var, na.rm = TRUE))
#>          x      x_se
#> 1 5.182796 0.2990772

# dplyr equivalent
data_srvyr$variables %>%
    summarize(x = mean(numeric_var))
#> # A tibble: 1 x 1
#>       x
#>   <dbl>
#> 1    NA

data_srvyr$variables %>%
    summarize(x = mean(numeric_var, na.rm = TRUE))
#> # A tibble: 1 x 1
#>       x
#>   <dbl>
#> 1  5.18



# Before this was fixed in 0.4.0, this would lead to an error because x was length 3, but y was length 4
data_srvyr %>%
    group_by(I.HB_LEARNING.ratio_schools) %>%
    summarize(
        x = survey_mean(),
        y = survey_mean(numeric_var, na.rm = TRUE)
    )
#> # A tibble: 4 x 5
#>   I.HB_LEARNING.ratio_schools     x   x_se     y  y_se
#>   <fct>                       <dbl>  <dbl> <dbl> <dbl>
#> 1 high                         0.43 0.0498  5.51 0.448
#> 2 ideal                        0.01 0.0100 10    0    
#> 3 low                          0.02 0.0141  4    1.42 
#> 4 <NA>                         0.54 0.0501  4.88 0.404
```

<sup>Created on 2020-11-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>
